crates/sui-framework/sources/external_resource.move

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Sui types for specifying off-chain/external resources.
+///
+/// The keywords "MUST", "MUST NOT", "SHOULD", "SHOULD NOT" and "MAY" below should be interpreted as described in
+/// RFC 2119.
+///
+/// TODO(ben): Should we s/resource/object? or will it just confuse people with Sui objects?
+///
+module sui::external_resource {
+    use std::ascii::{Self, String};
+    use std::vector;
+
+    /// Length of the vector<u8> representing a SHA256 digest.
+    const HASH_VECTOR_LENGTH: u64 = 32;
+
+    /// Error code when the length of the digest vector is not HASH_VECTOR_LENGTH.
+    const EHashLengthMismatch: u64 = 0;
+
+    /// URL: standard Uniform Resource Locator string.
+    ///
+    /// MUST follow RFC-3986.
+    /// Clients MUST support (at least) the following schemes: ipfs, https.
+    struct Url has store, copy, drop {
```

Comment:
I think this may want to continue living in a separate `url.move` file, both to avoid breaking changes for clients and for more idiomatic Move style (typically, each module should have at most one "main" type).